crates/notion-core/src/hook/tool.rs

```diff
@@ -0,0 +1,160 @@
+//! Types representing Notion Tool Hooks.
+
+use std::ffi::OsString;
+use std::io::Read;
+use std::process::{Command, Stdio};
+
+use path::{ARCH, OS};
+
+use cmdline_words_parser::StrExt;
+use notion_fail::{FailExt, Fallible, ResultExt};
+use semver::Version;
+
+const ARCH_TEMPLATE: &'static str = "{arch}";
+const OS_TEMPLATE: &'static str = "{os}";
+const VERSION_TEMPLATE: &'static str = "{version}";
+
+/// A Hook for resolving the distro URL for a given Tool Version
+#[derive(PartialEq, Debug)]
+pub enum DistroHook {
+    Prefix(String),
+    Template(String),
+    Bin(String),
+}
+
+impl DistroHook {
+    /// Performs resolution of the Distro URL based on the given
+    /// Version and File Name
+    pub fn resolve(&self, version: &Version, filename: &str) -> Fallible<String> {
+        match self {
+            &DistroHook::Prefix(ref prefix) => Ok(format!("{}{}", prefix, filename)),
+            &DistroHook::Template(ref template) => Ok(template
+                .replace(ARCH_TEMPLATE, ARCH)
```

Comment:
I think I'd want to see this templating logic get pulled out into a separate "simple templating" crate. I spent some time researching existing templating engines in the Rust ecosystem, and they all seem too feature-rich for our purposes here. I think what we want is something minimal. But we still probably want a _tiny_ bit of functionality, e.g. the ability to escape `'{'` characters just in case. So I think it'd be a little cleaner to pull this out into a crate (in-repo at least for now). I don't care deeply whether we pull it out now or in a followup issue.

Reply:
Filed #260 for pulling the behavior into a separate crate, as this PR is big enough already, and I think there are some outstanding questions for how we want the templating to work. To reduce the likelihood of needing an escape, however, I also updated the template strings to use `{{` and `}}`, instead of single braces.

Since it's currently using string replaces on a specific set of template strings, we don't have to worry about escaping arbitrary `{{` sequences. It will only be a problem if the user wants to have a literal `{{arch}}`, `{{os}}`, or `{{version}}` in their URL that they would need to escape.